Lazy Jinja2 context

uranusjr opened the pull request:
Fix #20213. To work around Jinja2 eagerly resolving context into a new dict, we implement a custom copy interface, and use lower level Jinja2 functions to render templates (because the higher level `Template.render()` simply calls `dict(**data)` which is not overridable).

Tests are added to ensure that

1. Rendering a template with deprecated variable emits a deprecation warning.
2. Rendering a template _without_ deprecated variable does not emit a deprecation warning.

uranusjr committed:
Make unexpected template depracation fail test

uranusjr committed:
Only show deprecation when template accesses vars

uranusjr committed:
Fix native type rendering

uranusjr committed:
AirflowContextDeprecationWarning errors globally

uranusjr committed:
Use custom template to render task log filename

uranusjr committed:
Only catch context deprecations in airflow proper

jedcunningham commented on airflow/utils/context.py:
```diff
@@ -124,6 +142,11 @@ def __reduce_ex__(self, protocol: int) -> Tuple[Any, ...]:
+
     def __getitem__(self, key: str) -> Any:
         with contextlib.suppress(KeyError):
             warnings.warn(_create_deprecation_warning(key, self._deprecation_replacements[key]), stacklevel=2)
```
```suggestion
            warnings.warn(_create_deprecation_warning(key, self._deprecation_replacements[key]))
```

This is the culprit, as it issues the warning from `jinja2/runtime.py:154` instead.

uranusjr replied:
Thanks! I’m not quite sure why Jinja2 seemingly made one call stack disappear, but it is what it is.

uranusjr committed:
Reduce deprecation warning stacklevel

Jinja2 makes the call stack in the context object disappear.
Not sure how it does it, but it does.

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
